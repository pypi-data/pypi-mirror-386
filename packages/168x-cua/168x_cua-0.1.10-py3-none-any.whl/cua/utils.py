import socket

def get_primary_ip() -> str:
    # doesn't send any packets; safe and fast
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    finally:
        s.close()