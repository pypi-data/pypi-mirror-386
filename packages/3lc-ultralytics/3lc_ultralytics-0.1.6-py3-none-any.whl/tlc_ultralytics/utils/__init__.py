from .check_requirements import check_requirements
from .dataset import check_tlc_dataset, parse_3lc_yaml_file
from .embeddings import reduce_embeddings
from .sampler import create_sampler
from .schemas import image_embeddings_schema, training_phase_schema

__all__ = (
    "check_requirements",
    "check_tlc_dataset",
    "create_sampler",
    "get_table_value_map",
    "image_embeddings_schema",
    "parse_3lc_yaml_file",
    "reduce_embeddings",
    "training_phase_schema",
)
