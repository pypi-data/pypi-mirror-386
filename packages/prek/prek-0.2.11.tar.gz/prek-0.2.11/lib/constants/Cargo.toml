[package]
name = "constants"
version = "0.0.1"
repository = { workspace = true }
edition = { workspace = true }
license = { workspace = true }

[dependencies]
tracing = { workspace = true }

[lints]
workspace = true
