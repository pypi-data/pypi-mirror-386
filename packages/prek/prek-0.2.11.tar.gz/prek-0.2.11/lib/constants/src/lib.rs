pub mod env_vars;

pub const CONFIG_FILE: &str = ".pre-commit-config.yaml";
pub const ALT_CONFIG_FILE: &str = ".pre-commit-config.yml";
pub const MANIFEST_FILE: &str = ".pre-commit-hooks.yaml";
