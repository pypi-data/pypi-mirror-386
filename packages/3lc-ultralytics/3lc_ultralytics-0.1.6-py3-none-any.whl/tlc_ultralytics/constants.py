import tlc
from ultralytics.utils import colorstr

# Column names
TRAINING_PHASE = "Training Phase"
IMAGE_COLUMN_NAME = tlc.IMAGE
CLASSIFY_LABEL_COLUMN_NAME = tlc.LABEL
DETECTION_LABEL_COLUMN_NAME = f"{tlc.BOUNDING_BOXES}.{tlc.BOUNDING_BOX_LIST}.{tlc.LABEL}"
SEGMENTATION_LABEL_COLUMN_NAME = f"{tlc.SEGMENTATIONS}.{tlc.INSTANCE_PROPERTIES}.{tlc.LABEL}"
OBB_LABEL_COLUMN_NAME = tlc.ORIENTED_BBS_2D
POSE_LABEL_COLUMN_NAME = tlc.KEYPOINTS_2D
PRECISION = "precision"
PRECISION_SEG = "precision_seg"
RECALL = "recall"
RECALL_SEG = "recall_seg"
MAP = "mAP"
MAP_SEG = "mAP_seg"
MAP50_95 = "mAP50-95"
MAP50_95_SEG = "mAP50-95_seg"
NUM_IMAGES = "num_images"
NUM_INSTANCES = "num_instances"
PER_CLASS_METRICS_STREAM_NAME = "per_class_metrics"

# Other
DEFAULT_TRAIN_RUN_DESCRIPTION = ""
DEFAULT_COLLECT_RUN_DESCRIPTION = "Created with model.collect()"

TLC_PREFIX = "3LC://"
TLC_COLORSTR = colorstr("3lc: ")

REQUIREMENTS_TO_CHECK = [
    ("3lc", "tlc"),
    ("ultralytics", "ultralytics"),
]
