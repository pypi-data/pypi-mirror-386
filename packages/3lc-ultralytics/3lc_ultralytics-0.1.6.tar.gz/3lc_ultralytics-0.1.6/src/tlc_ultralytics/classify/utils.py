from __future__ import annotations

import tlc
from ultralytics.data.utils import IMG_FORMATS
from ultralytics.utils import ROOT, YAML

from tlc_ultralytics.settings import Settings


def get_or_create_cls_table(
    key: str,
    data_dict: dict[str, object],
    image_column_name: str,
    label_column_name: str,
    project_name: str,
    dataset_name: str,
    table_name: str,
    settings: Settings | None = None,
) -> tlc.Table:
    """Get or create a classification table from a dataset dictionary.

    :param data_dict: Dictionary of dataset information
    :param project_name: Name of the project
    :param dataset_name: Name of the dataset
    :param table_name: Name of the table
    :param image_column_name: Name of the column containing image paths
    :param label_column_name: Name of the column containing labels
    :return: A tlc.Table.from_image_folder() table
    """

    # Special handling for ImageNet
    is_imagenet = (
        isinstance(data_dict["names"], dict)
        and isinstance(data_dict["names"][0], str)
        and data_dict["names"][0].startswith("n0")
    )

    if is_imagenet:
        label_overrides = YAML.load(ROOT / "cfg/datasets/ImageNet.yaml")["map"]
        label_overrides = {k: tlc.MapElement(internal_name=k, display_name=v) for k, v in label_overrides.items()}
    else:
        label_overrides = None

    return tlc.Table.from_image_folder(
        root=data_dict[key],
        image_column_name=image_column_name,
        label_column_name=label_column_name,
        project_name=project_name,
        dataset_name=dataset_name,
        table_name=table_name,
        extensions=IMG_FORMATS,
        if_exists="reuse",
        add_weight_column=True,
        description="Created with 3LC YOLO integration",
        label_overrides=label_overrides,
    )


def check_cls_table(table: tlc.Table, image_column_name: str, label_column_name: str) -> None:
    """Check that a table is compatible with the current task."""
    row_schema = table.row_schema.values

    try:
        # Check for image and label columns in schema
        assert image_column_name in row_schema, (
            f"Image column '{image_column_name}' not found in schema for Table {table.url}. "
            "Try providing your 'image_column_name' as an argument if you have a different column name."
        )
        assert label_column_name in row_schema, (
            f"Label column '{label_column_name}' not found in schema for Table {table.url}. "
            "Try providing your 'label_column_name' as an argument if you have a different column name."
        )

        # Check for desired roles
        assert row_schema[image_column_name].value.string_role == tlc.STRING_ROLE_IMAGE_URL, (
            f"Image column '{image_column_name}' must have role tlc.STRING_ROLE_IMAGE_URL={tlc.STRING_ROLE_IMAGE_URL}."
        )
        assert row_schema[label_column_name].value.number_role == tlc.LABEL, (
            f"Label column '{label_column_name}' must have role tlc.LABEL={tlc.LABEL}."
        )
    except AssertionError as e:
        raise ValueError(f"Table {table.url} is not compatible with YOLO classification.") from e
