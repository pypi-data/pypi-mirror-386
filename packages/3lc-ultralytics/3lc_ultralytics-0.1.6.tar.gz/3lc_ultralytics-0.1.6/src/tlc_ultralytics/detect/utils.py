from __future__ import annotations

import tlc
from tlc.client.torch.metrics.metrics_collectors.bounding_box_metrics_collector import (
    _TLCPredictedBoundingBox,
    _TLCPredictedBoundingBoxes,
)

from tlc_ultralytics.detect.dataset import TLCYOLODataset
from tlc_ultralytics.settings import Settings


def get_or_create_det_table(
    key: str,
    data_dict: dict[str, object],
    image_column_name: str,
    label_column_name: str,
    project_name: str,
    dataset_name: str,
    table_name: str,
    settings: Settings | None = None,
) -> tlc.Table:
    """Get or create a detection table from a dataset dictionary.

    :param key: The key of the dataset dictionary (the split to use)
    :param data_dict: Dictionary of dataset information
    :param project_name: Name of the project
    :param dataset_name: Name of the dataset
    :param table_name: Name of the table
    :param image_column_name: Name of the column containing image paths
    :param label_column_name: Name of the column containing labels
    :return: A tlc.Table.from_yolo() table
    """
    return tlc.Table.from_yolo(
        dataset_yaml_file=data_dict["yaml_file"],
        split=key,
        override_split_path=data_dict[key],
        task="detect",
        project_name=project_name,
        dataset_name=dataset_name,
        table_name=table_name,
        if_exists="reuse",
        add_weight_column=True,
        description="Created with 3LC YOLO integration",
    )


def build_tlc_yolo_dataset(
    cfg,
    table,
    batch,
    data,
    mode="train",
    rect=False,
    stride=32,
    multi_modal=False,
    exclude_zero=False,
    class_map=None,
    split=None,
    image_column_name=None,
    label_column_name=None,
):
    if multi_modal:
        return ValueError("Multi-modal datasets are not supported in the 3LC Ultralytics integration.")

    return TLCYOLODataset(
        table,
        exclude_zero=exclude_zero,
        class_map=class_map,
        imgsz=cfg.imgsz,
        batch_size=batch,
        augment=mode == "train",  # augmentation
        hyp=cfg,  # TODO: probably add a get_hyps_from_cfg function
        rect=cfg.rect or rect,  # rectangular batches
        cache=cfg.cache or None,
        single_cls=cfg.single_cls or False,
        stride=int(stride),
        pad=0.0 if mode == "train" else 0.5,
        prefix=split or mode,
        task=cfg.task,
        classes=cfg.classes,
        data=data,
        fraction=cfg.fraction if mode == "train" else 1.0,
        image_column_name=image_column_name,
        label_column_name=label_column_name,
    )


def check_det_table(
    table: tlc.Table,
    image_column_name: str = tlc.IMAGE,
    label_column_name: str = f"{tlc.BOUNDING_BOXES}.{tlc.BOUNDING_BOX_LIST}.{tlc.LABEL}",
) -> None:
    """Check that a table is compatible with the detection task in the 3LC YOLO integration.

    :param table: The table to check.
    :param image_column_name: The name of the column containing image paths.
    :param label_column_name: The full label path of the column containing labels.
    :raises: ValueError if the table is not compatible with the detection task.
    """
    row_schema = table.row_schema.values

    bounding_boxes_column_key, bounding_boxes_list_key, label_key = label_column_name.split(".")

    try:
        assert image_column_name in row_schema, f"Image column '{image_column_name}' not found."
        assert bounding_boxes_column_key in row_schema, f"Bounding box column '{bounding_boxes_column_key}' not found."
        assert bounding_boxes_list_key in row_schema[bounding_boxes_column_key].values, (
            f"Bounding box list '{bounding_boxes_list_key}' not found in column '{bounding_boxes_column_key}'."
        )

        assert tlc.IMAGE_HEIGHT in row_schema[bounding_boxes_column_key].values, (
            f"Bounding box column '{bounding_boxes_column_key}' does not contain a key '{tlc.IMAGE_HEIGHT}'."
        )
        assert tlc.IMAGE_WIDTH in row_schema[bounding_boxes_column_key].values, (
            f"Bounding box column '{bounding_boxes_column_key}' does not contain a key '{tlc.IMAGE_WIDTH}'."
        )

        for coordinate in [tlc.X0, tlc.Y0, tlc.X1, tlc.Y1]:
            assert coordinate in row_schema[bounding_boxes_column_key].values[bounding_boxes_list_key].values, (
                f"Bounding box list '{bounding_boxes_list_key}' in column '{bounding_boxes_column_key}' "
                f"does not contain a key '{coordinate}'."
            )
        assert label_key in row_schema[bounding_boxes_column_key].values[bounding_boxes_list_key].values, (
            f"Bounding box list '{bounding_boxes_list_key}' in column '{bounding_boxes_column_key}' "
            f"does not contain a key '{label_key}'."
        )
        assert table.get_value_map(label_column_name) is not None, (
            f"Unable to get value map for label value path {label_column_name}. Ensure that the table is compatible "
            "with the detection task or provide a `label_column_name` that matches the value path to the labels."
        )

    except (AssertionError, KeyError) as e:
        raise ValueError(f"Table with url {table.url} is not compatible with YOLO object detection. {e}") from None


def yolo_predicted_bounding_box_schema(
    label_value_map: dict[float, tlc.MapElement],
) -> tlc.Schema:
    """Create a 3LC bounding box schema for YOLO.

    :param categories: Categories for the current dataset.
    :returns: The YOLO bounding box schema for predicted boxes.
    """

    bounding_box_schema = tlc.BoundingBoxListSchema(
        label_value_map=label_value_map,
        x0_number_role=tlc.NUMBER_ROLE_BB_CENTER_X,
        x1_number_role=tlc.NUMBER_ROLE_BB_SIZE_X,
        y0_number_role=tlc.NUMBER_ROLE_BB_CENTER_Y,
        y1_number_role=tlc.NUMBER_ROLE_BB_SIZE_Y,
        x0_unit=tlc.UNIT_RELATIVE,
        y0_unit=tlc.UNIT_RELATIVE,
        x1_unit=tlc.UNIT_RELATIVE,
        y1_unit=tlc.UNIT_RELATIVE,
        description="Predicted Bounding Boxes",
        writable=False,
        is_prediction=True,
        include_segmentation=False,
    )

    return bounding_box_schema


def yolo_loss_schemas(training: bool = False) -> dict[str, tlc.Schema]:
    """Create a 3LC schema for YOLO per-sample loss metrics.

    :param training: Whether metrics are collected during training.
    :returns: The YOLO loss schemas for each of the three components.
    """
    schemas = {}
    schemas["box_loss"] = tlc.Schema(
        description="Box Loss",
        writable=False,
        value=tlc.Float32Value(),
        display_importance=3004,
    )
    schemas["dfl_loss"] = tlc.Schema(
        description="Distribution Focal Loss",
        writable=False,
        value=tlc.Float32Value(),
        display_importance=3005,
    )
    schemas["cls_loss"] = tlc.Schema(
        description="Classification Loss",
        writable=False,
        value=tlc.Float32Value(),
        display_importance=3006,
    )
    if training:
        schemas["loss"] = tlc.Schema(
            description="Weighted sum of box, DFL, and classification losses used in training",
            writable=False,
            value=tlc.Float32Value(),
            display_importance=3007,
        )
    return schemas


def construct_bbox_struct(
    predicted_annotations: list[dict[str, int | float | dict[str, float]]],
    image_width: int,
    image_height: int,
    inverse_label_mapping: dict[int, int] | None = None,
) -> _TLCPredictedBoundingBoxes:
    """Construct a 3LC bounding box struct from a list of bounding boxes.

    :param predicted_annotations: A list of predicted bounding boxes.
    :param image_width: The width of the image.
    :param image_height: The height of the image.
    :param inverse_label_mapping: A mapping from predicted label to category id.
    """

    bbox_struct = _TLCPredictedBoundingBoxes(
        bb_list=[],
        image_width=image_width,
        image_height=image_height,
    )

    for pred in predicted_annotations:
        bbox, label, score, iou = (
            pred["bbox"],
            pred["category_id"],
            pred["score"],
            pred["iou"],
        )
        label_val = inverse_label_mapping[label] if inverse_label_mapping is not None else label
        bbox_struct["bb_list"].append(
            _TLCPredictedBoundingBox(
                label=label_val,
                confidence=score,
                iou=iou,
                x0=bbox[0],
                y0=bbox[1],
                x1=bbox[2],
                y1=bbox[3],
            )
        )

    return bbox_struct
