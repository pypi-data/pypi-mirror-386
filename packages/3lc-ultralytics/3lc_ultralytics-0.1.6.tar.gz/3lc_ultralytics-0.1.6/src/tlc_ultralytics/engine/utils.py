from __future__ import annotations

import contextlib
import random

from ultralytics.utils import LOGGER

from tlc_ultralytics.constants import TLC_COLORSTR


def _complete_label_column_name(label_column_name: str, default_label_column_name: str) -> str:
    """Create a complete label column name from a potentially partial one.

    Examples:
        >>> _complete_label_column_name("a", "a")
        "a"
        >>> _complete_label_column_name("a", "a.b.c")
        "a.b.c"
        >>> _complete_label_column_name("a.b.c", "d.e.f")
        "a.b.c"
        >>> _complete_label_column_name("", "a.b.c")
        "a.b.c"
    """
    parts = label_column_name.split(".") if label_column_name else []
    default_parts = default_label_column_name.split(".")

    for i, default_part in enumerate(default_parts):
        if i >= len(parts):
            parts.append(default_part)

    return ".".join(parts)


@contextlib.contextmanager
def _restore_random_state():
    """Context manager to ensure the global random state is unchanged by the wrapped code."""
    state = random.getstate()
    yield
    random.setstate(state)


def _handle_deprecated_column_name(
    arg_value: str | None, settings_value: str | None, default_value: str, column_name: str
) -> str:
    """Handling for when a column name is passed as an argument directly, instead of through a `Settings` object.
    Used in the `trainer` and `validator` classes. A warning is logged if the column name is passed as an argument.

    If the column name is provided both through the `Settings` object and as an argument, the one passed directly is
    used and a warning is logged.

    If the column name is not provided, the default value is returned.

    :param arg_value: The column name passed as an argument directly.
    :param settings_value: The column name set in the `Settings` object.
    :param default_value: The default column name.
    :return: The column name.
    """
    if arg_value is not None:
        msg = (
            f"Passing `{column_name}` as an argument is deprecated. Provide `{column_name}` to a `Settings` object "
            "instead."
        )
        LOGGER.warning(f"{TLC_COLORSTR}{msg}")

        if settings_value is not None:
            msg = (
                f"`{column_name}` is both set in the `Settings` object and provided directly. Using the one passed "
                f"directly: '{arg_value}'."
            )
            LOGGER.warning(f"{TLC_COLORSTR}{msg}")

        return arg_value
    elif settings_value is None:
        return default_value
    return settings_value
