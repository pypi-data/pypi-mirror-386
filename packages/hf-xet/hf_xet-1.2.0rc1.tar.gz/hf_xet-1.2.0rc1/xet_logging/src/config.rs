use std::path::{Path, PathBuf};
use std::time::Duration;

use utils::normalized_path_from_user_string;

use crate::constants::*;

#[derive(Clone, Debug, PartialEq)]
pub enum LoggingMode {
    Directory(PathBuf),
    File(PathBuf),
    Console,
}

/// The log directory cleanup configuration.  By default, the values
/// are loaded from environment variables.

#[derive(Clone, Debug, PartialEq)]
pub struct LogDirConfig {
    pub min_deletion_age: Duration,
    pub max_retention_age: Duration,
    pub size_limit: u64,
    pub filename_prefix: String,
}

impl Default for LogDirConfig {
    fn default() -> Self {
        // Load the defaults from the environmental config.
        Self {
            min_deletion_age: *LOG_DIR_MIN_DELETION_AGE,
            max_retention_age: *LOG_DIR_MAX_RETENTION_AGE,
            size_limit: LOG_DIR_MAX_SIZE.as_u64(),
            filename_prefix: LOG_PREFIX.to_string(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct LoggingConfig {
    pub logging_mode: LoggingMode,
    pub use_json: bool,
    pub enable_log_dir_cleanup: bool,
    pub version: String,
    pub log_dir_config: LogDirConfig,
}

impl LoggingConfig {
    /// Set up logging to a directory.  Note that this can be overwritten by environmental
    pub fn default_to_directory(version: String, log_directory: impl AsRef<Path>) -> LoggingConfig {
        // Choose the logging mode.
        let logging_mode = {
            if let Some(log_dest) = &*LOG_DEST {
                if log_dest.is_empty() {
                    LoggingMode::Console
                } else {
                    let path = normalized_path_from_user_string(log_dest);

                    if log_dest.ends_with('/')
                        || (cfg!(windows) && log_dest.ends_with('\\'))
                        || (path.exists() && path.is_dir())
                    {
                        LoggingMode::Directory(path)
                    } else {
                        LoggingMode::File(path)
                    }
                }
            } else {
                LoggingMode::Directory(log_directory.as_ref().to_path_buf())
            }
        };

        let use_json = {
            if let Some(format) = &*LOG_FORMAT {
                format.to_ascii_lowercase().trim() == "json"
            } else {
                logging_mode != LoggingMode::Console
            }
        };

        let enable_log_dir_cleanup = matches!(logging_mode, LoggingMode::Directory(_)) && !*LOG_DIR_DISABLE_CLEANUP;

        Self {
            logging_mode,
            use_json,
            enable_log_dir_cleanup,
            version,
            log_dir_config: LogDirConfig::default(),
        }
    }
}
