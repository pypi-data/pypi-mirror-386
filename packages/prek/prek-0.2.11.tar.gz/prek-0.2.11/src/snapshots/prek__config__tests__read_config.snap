---
source: src/config.rs
expression: config
---
Config {
    repos: [
        Remote(
            RemoteRepo {
                repo: "https://github.com/abravalheri/validate-pyproject",
                rev: "v0.20.2",
                hooks: [
                    RemoteHook {
                        id: "validate-pyproject",
                        name: None,
                        entry: None,
                        language: None,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: None,
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: None,
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
        Remote(
            RemoteRepo {
                repo: "https://github.com/crate-ci/typos",
                rev: "v1.26.0",
                hooks: [
                    RemoteHook {
                        id: "typos",
                        name: None,
                        entry: None,
                        language: None,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: None,
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: None,
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
        Local(
            LocalRepo {
                hooks: [
                    ManifestHook {
                        id: "cargo-fmt",
                        name: "cargo fmt",
                        entry: "cargo fmt --",
                        language: System,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: Some(
                                [
                                    "rust",
                                ],
                            ),
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: Some(
                                false,
                            ),
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
        Local(
            LocalRepo {
                hooks: [
                    ManifestHook {
                        id: "cargo-dev-generate-all",
                        name: "cargo dev generate-all",
                        entry: "cargo dev generate-all",
                        language: System,
                        options: HookOptions {
                            alias: None,
                            files: Some(
                                SerdeRegex(
                                    "^crates/(uv-cli|uv-settings)/",
                                ),
                            ),
                            exclude: None,
                            types: Some(
                                [
                                    "rust",
                                ],
                            ),
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: Some(
                                false,
                            ),
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
        Remote(
            RemoteRepo {
                repo: "https://github.com/pre-commit/mirrors-prettier",
                rev: "v3.1.0",
                hooks: [
                    RemoteHook {
                        id: "prettier",
                        name: None,
                        entry: None,
                        language: None,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: None,
                            types_or: Some(
                                [
                                    "yaml",
                                    "json5",
                                ],
                            ),
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: None,
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
        Remote(
            RemoteRepo {
                repo: "https://github.com/astral-sh/ruff-pre-commit",
                rev: "v0.6.9",
                hooks: [
                    RemoteHook {
                        id: "ruff-format",
                        name: None,
                        entry: None,
                        language: None,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: None,
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: None,
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: None,
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                    RemoteHook {
                        id: "ruff",
                        name: None,
                        entry: None,
                        language: None,
                        options: HookOptions {
                            alias: None,
                            files: None,
                            exclude: None,
                            types: None,
                            types_or: None,
                            exclude_types: None,
                            additional_dependencies: None,
                            args: Some(
                                [
                                    "--fix",
                                    "--exit-non-zero-on-fix",
                                ],
                            ),
                            always_run: None,
                            fail_fast: None,
                            pass_filenames: None,
                            description: None,
                            language_version: None,
                            log_file: None,
                            require_serial: None,
                            stages: None,
                            verbose: None,
                            minimum_prek_version: None,
                        },
                    },
                ],
            },
        ),
    ],
    default_install_hook_types: None,
    default_language_version: None,
    default_stages: None,
    files: None,
    exclude: Some(
        SerdeRegex(
            "(?x)^(\n  .*/(snapshots)/.*|\n)$\n",
        ),
    ),
    fail_fast: Some(
        true,
    ),
    minimum_prek_version: None,
}
